include/ppx/metrics.h

```diff
@@ -0,0 +1,198 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ppx_metrics_h
+#define ppx_metrics_h
+
+#include "ppx/config.h"
+
+#include <limits>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace ppx {
+namespace metrics {
+
+#define METRICS_NO_COPY(type)   \
+    type(type&)       = delete; \
+    type(const type&) = delete; \
+    type& operator=(const type&) = delete;
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricInterpretation
+{
+    NONE,
+    HIGHER_IS_BETTER,
+    LOWER_IS_BETTER,
+};
+
+struct Range
+{
+    double lowerBound = std::numeric_limits<double>::min();
+    double upperBound = std::numeric_limits<double>::max();
+};
+
+struct MetricMetadata
+{
+    std::string          name;
+    std::string          unit;
+    MetricInterpretation interpretation = MetricInterpretation::NONE;
+    Range                expectedRange;
+};
+
+////////////////////////////////////////////////////////////////////////////////
+
+enum class MetricType
+{
+    GAUGE,
+    COUNTER,
+};
+
+class Metric
+{
+    friend class Run;
+
+public:
+    const MetricMetadata& GetMetadata() const;
+    MetricType            GetType() const;
```

Comment:
Is this needed?
From what I see, the only cases we check for type are to return the correct metric from the tables. This could be removed by storing each metric in their own table right?

Reply:
Yes, the metric super class is gone.